Associate descriptions with question text

muhammad-ammar opened the pull request:
### [TNL-5014](https://openedx.atlassian.net/browse/TNL-5014)

This PR includes the below changes:
- New Markdown formatter and <description> tag to add description for question.
- Add description ids on desired elements.
- Remove label or p tag with question text to avoid rendering the question twice.
- Updated default YAML templates

**Note:** Currently we have below limitations. These will be addressed once we make accessibility changes for individual problem types.
- Questions and Descriptions are only rendered in Checkboxes and MultipleChoice problems.

### [Sandbox](https://studio-tnl5014.sandbox.edx.org/)

## Testing
- [x] Unit.
- [x] Acceptance.

## Post-review
- [x] Squash commits into discrete sets of changes with descriptive commit messages.

## Reviewers
- [x] @edx/edx-accessibility  
- [x] @ehteshamkafeel 
- [x] @cahrens 

@muzaffaryousaf FYI

muhammad-ammar commented:
@edx/edx-accessibility @cahrens @ehteshamkafeel  This is ready for review.

@muzaffaryousaf FYI

estute commented:
jenkins run quality

muhammad-ammar commented:
jenkins run python

muhammad-ammar commented:
jenkins run quality

muhammad-ammar commented:
@edx/edx-accessibility @cahrens @ehteshamkafeel Can you review the changes.

cahrens commented:
@muhammad-ammar I've been waiting for an initial review to be done by someone on your team (especially since this is a fairly large PR).

muhammad-ammar commented:
@ehteshamkafeel Can you review the changes today?

ehteshamkafeel commented:
@muhammad-ammar with these changes the accessibility of capa problems, that we are not updating/making accessible yet, like e.g. filesubmission, will effect. As we are removing the `aria-label` attribute from these problem types.  We should keep these problems as they were before these changes.

ehteshamkafeel commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -855,17 +860,55 @@ def _preprocess_problem(self, tree):  # private
+                # Delete <p> tag containing question text otherwise question will be rendered twice
+                p_tag = response.xpath("preceding-sibling::p[text()='{}']".format(label))
+                if p_tag:
+                    tree.remove(p_tag[0])
```
can we move the remove logic to one place to make it more readable.

ehteshamkafeel commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -855,17 +860,55 @@ def _preprocess_problem(self, tree):  # private
+                # This is the case when we have a problem with multiple questions having OLX only
+                label = inputfields[0].attrib['label']
+                # Delete <p> tag containing question text otherwise question will be rendered twice
+                p_tag = response.xpath("preceding-sibling::p[text()='{}']".format(label))
```
will this xapth look for all preceding element or just immediate preceding element?

muhammad-ammar replied:
immediate preceding sibling

ehteshamkafeel commented on common/lib/capa/capa/templates/choicetext.html:
```diff
@@ -9,8 +9,8 @@
-    
-    <fieldset aria-label="${label}">
+
+    <fieldset>
```
why we didn't put `aria-describedby` here?

muhammad-ammar replied:
We don't desired/expected HTML for choicetext problems.

ehteshamkafeel commented on common/lib/capa/capa/templates/textline.html:
```diff
@@ -27,15 +27,13 @@
           %if value:
             ${value|h}
-          % else:
-            ${label}
```
we need to keep them as is if not suggested otherwise in reference templates.

ehteshamkafeel commented on common/lib/capa/capa/tests/test_capa_problem.py:
```diff
@@ -0,0 +1,163 @@
+        <problem>
+            <choiceresponse>
+                <label>Select the correct synonym of paranoid?</label>
+                <description>Only the paranoid survive.</description>
```
We should check with multiple descriptions as well.

ehteshamkafeel commented on common/lib/capa/capa/tests/test_capa_problem.py:
```diff
@@ -0,0 +1,163 @@
+        This is the case when we have a markdown problem with multiple-questions.
+        In this case when markdown is converted to xml, there will be no label
+        tag and label attribute inside responsetype. But we have a label tag
+        before the responsetype.
```
But we can have label attribute outside responsetype in some `<p>`?

muhammad-ammar replied:
label attribute is always set on inputtype and inputtype always come inside responsetype so we can't have label attribute outside responsetype.

ehteshamkafeel commented on common/lib/capa/capa/tests/test_input_templates.py:
```diff
@@ -473,6 +503,19 @@ def test_message(self):
         xpath = "//span[@class='message']"
         self.assert_has_text(xml, xpath, self.context['msg'])
 
+    def test_description(self):
```
These tests are basically checking accessibility. Shouldn't they be marked as a11y tests?

muhammad-ammar replied:
Sorry, i didn't got you?

ehteshamkafeel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -149,11 +149,8 @@ div.problem {
-    font: inherit;
-    color: inherit;
+  .question-description {
+    @include margin(($baseline*0.75), 0, ($baseline*0.75), 0);
```
two if the margins are same. we should be able to just write `margin(($baseline*0.75), 0);`

ehteshamkafeel commented on common/lib/xmodule/xmodule/js/src/problem/edit.coffee:
```diff
@@ -209,6 +209,20 @@ class @MarkdownEditingDescriptor extends XModule.Descriptor
+              label = '<label>' + result[0] + '</label>' + '\n';
+
+          // don't add empty <description> tag
+          if (result.length == 1 || result[1].length == 0) {
```
Here result[1] will be a string so we can just check 
`if (result.length == 1 || !result[1])`

ehteshamkafeel commented:
@muhammad-ammar Other than few comments it's looking good to me. Tag me again when you are done with conditional removal of label.

muhammad-ammar commented:
@ehteshamkafeel Thanks for the review. I have addressed your feedback.

muhammad-ammar commented:
@cahrens Can you have a look at the changes.

muhammad-ammar commented:
@edx/edx-accessibility Can you do an accessibility review?

cahrens commented:
Starting to look at this. In the existing problem on this unit (https://tnl5014.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/courseware/interactive_demonstrations/basic_questions/) there are empty aria-describedby attributes. I don't know how this problem was authored; creating a new multiple choice problem seems to work OK. Is it expected that we will see empty aria-desribedby attributes for pre-existing problems?

cahrens commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -855,17 +870,67 @@ def _preprocess_problem(self, tree):  # private
+            element_to_be_deleted = None
+
+            # Extract label value from <label> tag or label attribute from inside the responsetype
+            # Extract label value from <label> tag
```
Duplicated line

cahrens commented:
@muzaffaryousaf I'm having trouble remembering the details of the decision we made on the whiteboard a few weeks back (regarding avoiding duplication of the label). Can you send me a summary of the decision or point me to where that is recorded?

cahrens commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -855,17 +870,67 @@ def _preprocess_problem(self, tree):  # private
+                    element_to_be_deleted = p_tag[0]
+            else:
+                # neither <label> tag nor label attribute is present inside responsetype
+                # existing problem with multi-questions without --- having markdown
```
I'm having trouble understanding this comment. So do you know for sure that this is an existing problem with multi-questions and no"---"? How do you know hat?

muhammad-ammar replied:
@cahrens This can happen whether the problem is new or existing if the problem has multiple-questions not separated according to new format. I will update the comments.

cahrens commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -855,17 +870,67 @@ def _preprocess_problem(self, tree):  # private
+            # for non-accessible responsetypes it may be possible that label attribute is not present
+            # in this case pass an empty label. remember label attribute is only used as value for aria-label
+            if response.tag not in ACCESSIBLE_CAPA_RESPONSE_TYPES and label == DEFAULT_QUESTION_TEXT:
+                label = ''
```
Will the empty label cause an accessibility failure (is this the case I saw with the empty aria-describedby)?

muhammad-ammar replied:
@cahrens Question label will never be empty for fully accessible problems. We will set it to default value if its empty. The above check is only used for problems which are not fully-accessible.

ehteshamkafeel replied:
@muhammad-ammar can't we leave the label as is in this case? i.e. DEFAULT_QUESTION_TEXT?

cahrens commented on common/lib/capa/capa/templates/optioninput.html:
```diff
@@ -1,7 +1,7 @@
 
 <form class="inputtype option-input ${doinline}">
-   <select name="input_${id}" id="input_${id}" aria-label="${label}" aria-describedby="answer_${id}">
+   <select name="input_${id}" id="input_${id}" aria-label="${response_data['label']}" aria-describedby="answer_${id} ${response_data['description_ids']}">
```
Why does this template use both response_data['label'] and response_data['description_ids']? It looks like the other templates use one or the other (I think based on whether or not they are "fully accessible").

muhammad-ammar replied:
@cahrens `aria-label="${response_data['label']}"` will be removed in [TNL-4970](https://openedx.atlassian.net/browse/TNL-4970). Didn't removed in this PR to avoid test failures.

cahrens commented on common/lib/capa/capa/templates/textline.html:
```diff
@@ -14,7 +14,7 @@
     % endif
 
-  <input type="text" name="input_${id}" id="input_${id}" aria-label="${label}" aria-describedby="answer_${id}" value="${value|h}"
+  <input type="text" name="input_${id}" id="input_${id}" aria-label="${response_data['label']}" aria-describedby="answer_${id} ${response_data['description_ids']}" value="${value|h}"
```
Same question here.

muhammad-ammar replied:
@cahrens `aria-label="${response_data['label']}"` will be removed in https://github.com/edx/edx-platform/pull/13121. Didn't removed in this PR to avoid test failures.

cahrens commented on common/lib/capa/capa/tests/test_capa_problem.py:
```diff
@@ -0,0 +1,220 @@
+        * <label> tag is removed to avoid duplication
+
+        This is the case when we have a problem with single question or
+        problem with multiple-questions separated as per the new format.
```
Is it worth also having a test case with multiple question text (separated per the new format)?

muhammad-ammar replied:
@cahrens added a test.

cahrens commented on common/lib/capa/capa/tests/test_capa_problem.py:
```diff
@@ -0,0 +1,220 @@
+        Verify that label is extracted and <p> tag with question
+        text is removed when label attribute is set on inputtype.
+
+        This is the case when we have a OLX only problem with multiple-questions.
```
Why does the XML below only have a single question?

muhammad-ammar replied:
@cahrens Comment needs to be updated.

cahrens commented on common/lib/capa/capa/tests/test_capa_problem.py:
```diff
@@ -0,0 +1,220 @@
+        """
+        Verify that label is extracted correctly.
+
+        This is the case when we have a markdown problem with multiple-questions.
```
Again, the XML below does not have multiple questions. Can you clarify or change the XML to have multiple questions?

muhammad-ammar replied:
@cahrens Updated the test.

cahrens commented on common/lib/capa/capa/tests/test_capa_problem.py:
```diff
@@ -0,0 +1,220 @@
+
+    def test_default_question_text(self):
+        """
+        Verify that default question text is shown when we question is missing.
```
typo: "we question is missing"

cahrens commented on common/lib/capa/capa/tests/test_input_templates.py:
```diff
@@ -502,6 +545,19 @@ def test_size(self):
+        xml = self.render_to_xml(self.context)
+
+        # Verify that input aria-describedby attribute has correct description ids
+        fieldset_aria_describedby = xml.xpath("string(//input/@aria-describedby)")
```
Create a helper method that these tests can call? It looks like the only difference is the CSS selector.

cahrens commented on common/lib/xmodule/xmodule/js/src/problem/edit.coffee:
```diff
@@ -552,20 +537,20 @@ class @MarkdownEditingDescriptor extends XModule.Descriptor
-      splits = xml.split(/(\<\/?(?:script|pre).*?\>)/g);
-      scriptFlag = false;
+      splits = xml.split(/(\<\/?(?:script|pre|label|description).*?\>)/g);
+      wrappable = true;
```
Can you add a comment about what "wrappable" signifies? I'm having trouble figuring that out.

cahrens commented:
@muzaffaryousaf Done with my first pass.

muhammad-ammar commented:
> Is it expected that we will see empty aria-desribedby attributes for pre-existing problems?

@cahrens if descriptions are missing then `aria-desribedby` will be empty.

cahrens commented:
@muhammad-ammar Regarding the empty aria-describedby, can you make the template smarter and not generate it if there are no descriptions? An empty aria-describedby is an a11y error.

muhammad-ammar commented:
@cahrens I have addressed your feedback. Regarding empty `aria-describedby` i have checked `aria-describedby=""` in a bokchoy a11y test and test passed. I will further check it tomorrow.

cahrens commented:
@muhammad-ammar When I used browser a11y checks manually, I saw the failure for the empty aria-describedby. So regardless of whether or not the bok choy test finds it, it should still be fixed.

![image](https://cloud.githubusercontent.com/assets/484484/17297955/f11819cc-57d5-11e6-8b3e-7a4bdc5c10f6.png)

muhammad-ammar commented:
I will fix it. In the mean time can you check the addressed feedback.

On Mon, Aug 1, 2016, 7:51 PM Christina Roberts <notifications@github.com>
wrote:

> @muhammad-ammar <https://github.com/muhammad-ammar> When I used browser
> a11y checks manually, I saw the failure for the empty aria-describedby. So
> regardless of whether or not the bok choy test finds it, it should still be
> fixed.
>
> [image: image]
> <https://cloud.githubusercontent.com/assets/484484/17297955/f11819cc-57d5-11e6-8b3e-7a4bdc5c10f6.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/edx-platform/pull/13040#issuecomment-236603866>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGdFNIbWS59Q0sya-8q3I4Mzab-T27rkks5qbggOgaJpZM4JQyl7>
> .
>

cahrens commented:
Sure, I'll take a look today.

cahrens commented:
@muhammad-ammar @muzaffaryousaf it would be really helpful to me (and I imagine to @cptvitamin) if you can provide a description of the technical approach implemented by this PR (both for old-style and new-style problems). There were a lot of different iterations, and although I was present at the whiteboard meeting where the final approach was discussed, I don't recall the details.

cahrens commented on common/lib/xmodule/xmodule/js/src/problem/edit.coffee:
```diff
@@ -552,20 +537,22 @@ class @MarkdownEditingDescriptor extends XModule.Descriptor
-      scriptFlag = false;
+      splits = xml.split(/(\<\/?(?:script|pre|label|description).*?\>)/g);
+
+      // used to check if a text string be wrapped by <p> tag  
```
Typo in this comment.

And I'm still unsure what the conditions are in which a text string can/should be wrapped by a p tag. In words, when should wrappable be true vs. false?

cahrens commented on common/lib/xmodule/xmodule/js/src/problem/edit.coffee:
```diff
@@ -552,20 +537,22 @@ class @MarkdownEditingDescriptor extends XModule.Descriptor
-          if(/\<\/(script|pre)/.test(splits[i])) {
-              scriptFlag = false;
+          if(/\<\/(script|pre|label|description)/.test(splits[i])) {
+              wrappable = true;
```
How does this condition differ from lines 546-547 (where wrappable is the opposite)?

muhammad-ammar replied:
@cahrens This regex condition checks the end of tag whereas the previous regex condition checks the start of the tag.

cahrens commented:
I looked over the feedback commits and just had a few comments (above). But please do still provide the description of the technical approach.

muzaffaryousaf commented:
@cahrens To avoid the duplicate question text in case of legacy problems we will compare the value of label attribute with preceding paragraph tag (p tag before the responseType) and will ignore the **p** tag at the time of html rendering. If both values matches then we will pick the value from attribute otherwise we will pick the value from paragraph. 

Here is some pseudo code:
if label attribute == preceding paragraph, do not render paragraph and consider attribute as question title
else
consider the paragraph value as question text but do not render paragraph.

Let me know if it's not clear.

cahrens commented:
@muzaffaryousaf Thanks. And what is the algorithm for "new problems"? And how do you determine a "new problem" vs. a legacy one?

muzaffaryousaf commented:
@cahrens We planned to remove the label attribute functionality from the conversion code, it means that we won't have label attribute anymore except for the problems that are in XML only. 

Here is the algorithm to choose the question title:
1 - If we have label element, we'll consider that content as question title and won't render the label itself to avoid the duplication.

2 - If we don't have label element, we'll get the question title from label attribute, this is case for the legacy problems. This case has 2 scenarios which i mentioned above. 

Let me know if you need anything else to complete review for this PR. Thanks.

cptvitamin commented on common/lib/xmodule/xmodule/templates/problem/checkboxes_response_hint.yaml:
```diff
@@ -13,7 +13,7 @@ metadata:
       Use the following example problem as a model.
 
-      >>Which of the following is a fruit? Check all that apply.<<
+      >>Which of the following is a fruit? Check all that apply.||Make sure you select all of the correct options—there may be more than one!<<
```
I don't think we need the text "Check all that apply" here since we specify that in the description text.

cptvitamin commented on common/lib/xmodule/xmodule/templates/problem/checkboxes_response_hint.yaml:
```diff
@@ -37,7 +37,8 @@ data:   |
                 <p>Be sure to select Settings to specify a Display Name and other values that apply.</p>
                 <p>Use the following example problem as a model.</p>
                 <label>Which of the following is a fruit? Check all that apply.</label>
```
Remove "Check all that apply"

cptvitamin commented:
This looks great, minus that small comment about the template text.  I only tested functionality on newly authored Checkboxes and Radiobuttons.  Should I be looking at legacy problems?  I took a quick look at existing problems in the demo course and everything seemed to be OK, but I wasn't sure how to properly test that.

cahrens commented on common/lib/xmodule/xmodule/js/src/problem/edit.coffee:
```diff
@@ -552,20 +537,23 @@ class @MarkdownEditingDescriptor extends XModule.Descriptor
+
+      // Wrap a string by <p> tag when line is not already wrapped by another tag
+      // true when line is not already wrapped by another tag false otherwise
+      makeParagraph = true;
```
Thanks, this is much clearer!

cahrens commented:
👍

muhammad-ammar commented:
@cahrens Thanks for the review.

muzaffaryousaf commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -35,6 +37,16 @@
 # extra things displayed after "show answers" is pressed
 solution_tags = ['solution']
 
+# fully accesible capa response types
```
Typo --> `accessible`.

muzaffaryousaf commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -855,17 +870,76 @@ def _preprocess_problem(self, tree):  # private
+            # Extract label value from <label> tag or label attribute from inside the responsetype
+            responsetype_label_tag = response.find('label')
+            if responsetype_label_tag is not None:
+                label = responsetype_label_tag.text
```
Looks like the value of label would be question title, can we rename `label` to `question_title` or something similar ?

muzaffaryousaf commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -855,17 +870,76 @@ def _preprocess_problem(self, tree):  # private
+                # Delete <label> tag containing question text otherwise question will be rendered twice
+                element_to_be_deleted = responsetype_label_tag
+            elif 'label' in inputfields[0].attrib:
+                # Extract label value from label attribute
```
Extract question title from label attribute ?

muzaffaryousaf commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -855,17 +870,76 @@ def _preprocess_problem(self, tree):  # private
+                # * with multiple questions without separation
+                # * single question with old XML format only
+
+                label = inputfields[0].attrib['label']
```
same for `label` variable name.

muzaffaryousaf commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -855,17 +870,76 @@ def _preprocess_problem(self, tree):  # private
+                # * single question with old XML format only
+
+                label = inputfields[0].attrib['label']
+                # Get first <p> tag before responsetype
```
Can we add more explanation in comments about this p tag what is this used for.

muzaffaryousaf commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -855,17 +870,76 @@ def _preprocess_problem(self, tree):  # private
+
+                label = inputfields[0].attrib['label']
+                # Get first <p> tag before responsetype
+                p_tag = response.xpath('preceding-sibling::p[1]')
```
Can we compare the value of label attribute with all the preceding p tags because the author can put extra text between question title and responsetype and picking up 1st preceding p tag could lead us to show wrong question text.
@cptvitamin Do you know if authors are using extra text between the question title and response/input type in existing problems. If not then we can leave this as it.

muzaffaryousaf replied:
We decided to go with existing approach, so no action needed on this comment.

muhammad-ammar commented:
jenkins run bokchoy

muzaffaryousaf commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -855,17 +870,76 @@ def _preprocess_problem(self, tree):  # private
+            else:
+                # neither <label> tag nor label attribute is present inside responsetype
+                # This is the case when we have a problem with multi-questions without --- having markdown
+                # find the immediate <label> tag before the responsetype. also delete it avoid rendering twice
```
Can we update these comments with the following text

> In this case the problems don't have <label> tag or label attribute inside the responsetype so we will get the first preceding label tag w.r.t to this responsetype.
> This will take care of those multi-question problems that are not using `---` in their markdown.

muzaffaryousaf commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -855,17 +870,76 @@ def _preprocess_problem(self, tree):  # private
+            responsetype_label_tag = response.find('label')
+            if responsetype_label_tag is not None:
+                label = responsetype_label_tag.text
+                # Delete <label> tag containing question text otherwise question will be rendered twice
```
We're not deleting the element here but just picking it up to delete it later. Can we update this comment to show the actual behavior.

muzaffaryousaf commented:
@lamagnifica @sstack22 FYI, here is the PR about adding the question descriptions.

muzaffaryousaf commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -62,6 +74,9 @@
 # main class for this module
 
 
+DEFAULT_QUESTION_TEXT = _("You must specify meaningful question text.")
```
@lamagnifica This is the default text that we will show for every problem, in case if user does not provide question text or if they don't use the proper formatter to identify the title. Do you think this make sense ?

lamagnifica replied:
I conferred with several others, and we propose:
"Formatting error: You have not explicitly identified the question text." 

If it is possible to include a link that renders correctly (I know that has been problematic in the simple editor) it would be nice to have a link to 

http://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/course_components/create_problem.html#editing-a-problem-in-studio

worded something like "See the _edX documentation_ for more information.

cahrens replied:
We can't have links in that will work (at all) with the simple editor. Also, don't put "edX" into the text.

muzaffaryousaf replied:
For now we're going with text only. @muhammad-ammar Can you use this suggested text ?

muzaffaryousaf replied:
We also need to change the translation mechanism here.

muzaffaryousaf commented:
Since you already removed the label attributes from many places, can you also remove it from these places and replace them with label element.
- common/lib/capa/capa/responsetypes.py
- common/lib/capa/capa/tests/test_capa_problem.py
- common/lib/capa/capa/tests/test_inputtypes.py
- common/lib/xmodule/xmodule/tests/test_capa_module.py
- common/test/acceptance/tests/lms/test_library.py
- common/test/acceptance/tests/lms/test_lms_gating.py
- common/test/acceptance/tests/studio/test_studio_library_container.py
- common/test/acceptance/tests/xblock/test_crowdsourcehinter_problem.py
- lms/djangoapps/courseware/tests/test_transformers.py

We can also do this after we merge the PR into it's parent.

muzaffaryousaf commented on common/test/acceptance/tests/lms/test_lms_problems.py:
```diff
@@ -412,3 +416,94 @@ def test_logout_cancel_no_redirect(self):
+        """
+        Scenario: Test that question and description are rendered as expected.
+        Given I am enrolled in a course.
+        And I visit a unit page with a CAPA question.
```
`And` --> `When'

muzaffaryousaf commented on common/test/acceptance/tests/lms/test_lms_problems.py:
```diff
@@ -412,3 +416,94 @@ def test_logout_cancel_no_redirect(self):
+        """)
+        return XBlockFixtureDesc('problem', 'Problem A11Y TEST', data=xml)
+
+    def test_unique_ids(self):
```
How we're verifying that the ids are unique, test will fail if we don't have unique ids ?

muzaffaryousaf commented on common/test/acceptance/tests/lms/test_lms_problems.py:
```diff
@@ -412,3 +416,94 @@ def test_logout_cancel_no_redirect(self):
+        )
+
+        # Run the accessibility audit.
+        problem_page.a11y_audit.check_for_accessibility_errors()
```
@muhammad-ammar I think we don't need this test because we already have `common.test.acceptance.tests.lms.test_problem_types.ProblemTypeTestMixin#test_problem_type_a11y` every problem type.

ehteshamkafeel replied:
@muhammad-ammar why is this test necessary?

muzaffaryousaf commented on common/lib/capa/capa/templates/formulaequationinput.html:
```diff
@@ -3,8 +3,7 @@
         data-input-id="${id}" value="${value|h}"
-        aria-label="${label}"
-        aria-describedby="${id}_status"
+        ${describedby}
```
@muhammad-ammar Why do we need these changes and why we don't need `aria-label` ?

muzaffaryousaf commented on common/lib/capa/capa/tests/test_capa_problem.py:
```diff
@@ -0,0 +1,329 @@
+            <p>Be sure to check your spelling.</p>
+            <p>{}</p>
+            <stringresponse answer="vulnerable" type="ci">
+                <textline label="{}" size="40"/>
```
I see.. This label attribute does make sense, please ignore such existence from the list of files.

muzaffaryousaf commented on common/lib/capa/capa/tests/test_capa_problem.py:
```diff
@@ -0,0 +1,329 @@
+                '1_2':
+                {
+                    'description_ids': '',
+                    'label': 'Click the country which is home to the Pyramids.',
```
Can we use question variable here ?

muzaffaryousaf commented on common/lib/capa/capa/tests/test_capa_problem.py:
```diff
@@ -0,0 +1,329 @@
+        self.assertEqual(
+            problem.problem_data,
+            {
+                '1_2':
```
@muhammad-ammar Can you help me memorize why it's being started from 1_2 not 1_1 ?

muzaffaryousaf commented on common/lib/capa/capa/tests/test_capa_problem.py:
```diff
@@ -0,0 +1,329 @@
+        self.assertEqual(
+            len(problem.tree.xpath('//p[text()="{}"]'.format(question))),
+            0
+        )
```
Really liked the coverage that you made here. 💯

muzaffaryousaf commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -855,17 +870,76 @@ def _preprocess_problem(self, tree):  # private
-            responsetype_label = response.find('label')
-            problem_data[self.problem_id + '_' + str(response_id)] = {
-                'label': responsetype_label.text if responsetype_label is not None else ''
+            question_id = self.problem_id + '_' + str(response_id)
```
@muhammad-ammar Can we use format method here instead ?

muzaffaryousaf commented on common/lib/capa/capa/inputtypes.py:
```diff
@@ -319,8 +319,14 @@ def _get_render_context(self):
 
+        if self.response_data.get('description_ids'):
+            context.update(
+                {'describedby': 'aria-describedby="{}"'.format(self.response_data.get('description_ids'))}
```
Please add comment about the purpose of this change.

ehteshamkafeel replied:
@muhammad-ammar if we are passing the description_ids as a value of `aria-describedby` then we don't need to maintain an extra attribute in `response_data `

muzaffaryousaf commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -855,17 +870,76 @@ def _preprocess_problem(self, tree):  # private
+
+            problem_data[question_id] = {
+                'label': label,
+                'description_ids': ' '.join(descriptions.keys()),
```
Can we use keys of `descriptions` wherever we're using `description_ids`, this way we don't need to maintain this separate description_ids.

muzaffaryousaf commented:
@muhammad-ammar I'm done with the review, this looks good. I've left couple of minor comments, let me know when you're done with addressing them. Also, please update sandbox when you're done with changes.
Thanks.

muzaffaryousaf commented:
@ehteshamkafeel Can you finish you review on this ASAP ? We need to merge this today.

ehteshamkafeel commented on common/lib/capa/capa/templates/choicegroup.html:
```diff
@@ -33,8 +32,9 @@
-                class="${label_class}" >
-
+                class="${label_class}"
+                ${describedby}
```
@muhammad-ammar this is a bit confusing as by at looking at it one can't determine what `${describedby}` is if would be like `describedby=${description_ids}` it is much more understandable.

ehteshamkafeel commented on common/lib/capa/capa/templates/choicetext.html:
```diff
@@ -9,8 +9,8 @@
+
+    <fieldset aria-label="${response_data['label']}">
         % for choice_id, choice_description in choices:
         <%choice_id= choice_id %>
```
spacing issue, should be like
`<% choice_id = choice_id %>`

ehteshamkafeel replied:
oh sorry that's not your change.

ehteshamkafeel commented on common/lib/capa/capa/tests/test_capa_problem.py:
```diff
@@ -0,0 +1,329 @@
+        mismatched with question tag value.
+
+        This is the case when author updated the question <p> tag directly in XML but
+        didn't changed the label attribute value. In this case we will consider the
```
didn't change

ehteshamkafeel commented on common/lib/capa/capa/tests/test_input_templates.py:
```diff
@@ -112,6 +122,49 @@ def assert_has_text(self, xml_root, xpath, text, exact=True):
+
+        # for each xpath verify that description_ids are set correctly
+        for describedby_xpath in describedby_xpaths:
+            describedbys = xml.xpath(describedby_xpath)
```
Nit: describedby_ids

ehteshamkafeel commented on common/lib/xmodule/xmodule/js/src/problem/edit.coffee:
```diff
@@ -209,6 +209,20 @@ class @MarkdownEditingDescriptor extends XModule.Descriptor
+              label = '<label>' + result[0] + '</label>' + '\n';
+
+          // don't add empty <description> tag
+          if (result.length == 1 || !result[1]) {
```
use triple equal `===`

ehteshamkafeel commented on common/lib/xmodule/xmodule/js/src/problem/edit.coffee:
```diff
@@ -552,20 +537,23 @@ class @MarkdownEditingDescriptor extends XModule.Descriptor
       for (i = 0; i < splits.length; i += 1) {
-          if(/\<(script|pre)/.test(splits[i])) {
-              scriptFlag = true;
+          if(/\<(script|pre|label|description)/.test(splits[i])) {
```
formatting: space `if (`

ehteshamkafeel commented on common/lib/xmodule/xmodule/js/src/problem/edit.coffee:
```diff
@@ -552,20 +537,23 @@ class @MarkdownEditingDescriptor extends XModule.Descriptor
           }
 
-          if(!scriptFlag) {
+          if(makeParagraph) {
```
format properly.

ehteshamkafeel commented on common/lib/xmodule/xmodule/js/src/problem/edit.coffee:
```diff
@@ -552,20 +537,23 @@ class @MarkdownEditingDescriptor extends XModule.Descriptor
 
-          if(/\<\/(script|pre)/.test(splits[i])) {
-              scriptFlag = false;
+          if(/\<\/(script|pre|label|description)/.test(splits[i])) {
```
fix formatting.

ehteshamkafeel commented on common/lib/xmodule/xmodule/templates/problem/optionresponse_hint.yaml:
```diff
@@ -11,7 +11,7 @@ metadata:
         Use the following example problem as a model.
 
-        >> A/an ________ is a vegetable.<<
+        >> A/an ________ is a vegetable.||You can select only one option.<<
```
I think there should be better descriptions here. I think @lamagnifica can suggest some.

lamagnifica replied:
Please see the wiki pages! The last two comments on this page, https://openedx.atlassian.net/wiki/display/A11Y/Text+Input provide templates for text input and, separately, for text input with hints and feedback. Please let me know if you notice any problems -- there may be some cut and paste errors(!)
(I believe that replacing these examples with templates is scheduled as a separate task, though.)

muzaffaryousaf replied:
@ehteshamkafeel the templates are going to be change, so we can leave them as it

ehteshamkafeel commented:
@muhammad-ammar excellent work!! I have few comments otherwise it's looking in good shape.

muzaffaryousaf commented:
jenkins run all

muzaffaryousaf commented on common/lib/capa/capa/capa_problem.py:
```diff
@@ -855,17 +868,77 @@ def _preprocess_problem(self, tree):  # private
+                    label = label_tag[0].text
+                    element_to_be_deleted = label_tag[0]
+
+            _ = self.capa_system.i18n.ugettext
```
Do we need this anymore ?

muzaffaryousaf commented on common/lib/capa/capa/templates/choicegroup.html:
```diff
@@ -5,12 +5,11 @@
-        <legend id="${id}-legend" class="response-fieldset-legend question-text">${response_data['label']}</legend>
-      % else:
-        <legend>Question</legend>
-      % endif
```
@muhammad-ammar Don't we need this condition anymore ?
Also, don't we need `question-text`  as legend class ?

muzaffaryousaf commented:
👍 @muhammad-ammar LGTM. Don't forget to squash the commits before merge.

muzaffaryousaf commented:
jenkins run python
